Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -0,0 +1,106 @@
+/// This is the safest way for pipelining-unaware code to operate, as it ensures that
+/// mutation of any shared server state is not parallelised, and that responses are always
+/// sent for each request in turn. In almost all cases this is the behaviour that a
+/// pipeline will want.
+public class HTTPServerPipelineHandler: ChannelInboundHandler, ChannelOutboundHandler {
+    public typealias InboundIn = HTTPServerRequestPart
+    public typealias InboundOut = HTTPServerRequestPart
+    public typealias OutboundIn = HTTPServerResponsePart
+    public typealias OutboundOut = HTTPServerResponsePart
+
+    public init() { }
+
+    /// Whether the channel pipeline is currently processing a response.
+    private var responsePending = false
+
+    /// While we're waiting to send the response we don't read from the socket.
+    /// This keeps track of whether we need to call read() when we've send our response.
+    private var readPending = false
+
+    /// The buffered HTTP requests that are not going to be addressed yet.
+    private var requestBuffer: [NIOAny] = []
+
+    public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        if responsePending {
+            self.requestBuffer.append(data)
+            return
+        }
+
+        if case .end = self.unwrapInboundIn(data) {
+            // New request is complete. We don't want any more data from now on.
+            self.responsePending = true
+        }
+        ctx.fireChannelRead(data)
+    }
+
+    public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        precondition(self.responsePending, "Unexpected data flow")
+        var startReadingAgain = false
+        if case .end = self.unwrapOutboundIn(data) {
+            startReadingAgain = true
+        }
+
+        ctx.write(data, promise: promise)
+
+        if startReadingAgain {
+            self.deliverPendingRequests(ctx: ctx)
+            self.startReading(ctx: ctx)
+        }
+    }
+
+    public func read(ctx: ChannelHandlerContext) {
+        if self.responsePending {
+            self.readPending = true
+        } else {
+            ctx.read()
+        }
+    }
+
+    /// A response has been sent: we can now start passing reads through
+    /// again, and send any read() call we may have swallowed.
+    private func startReading(ctx: ChannelHandlerContext) {
+        if self.readPending {
+            self.readPending = false
+            ctx.read()
+        }
+    }
+
+    /// A response has been sent: deliver all pending requests and
+    /// mark the channel ready to handle more requests.
+    private func deliverPendingRequests(ctx: ChannelHandlerContext) {
+        self.responsePending = false
+        if self.requestBuffer.count > 0 {
+            let reads = self.requestBuffer
+            self.requestBuffer = []
+
+            reads.forEach { self.channelRead(ctx: ctx, data: $0) }
```

Comment:
See above ... If we switch to `CircularBuffer` we can just empty it and there is nothing left.